system/lib/pthread/library_pthread.c

```diff
@@ -420,6 +421,16 @@ EMSCRIPTEN_RESULT emscripten_wait_for_call_v(em_queued_call* call, double timeou
     while (!done && now < waitEndTime) {
       r = emscripten_futex_wait(&call->operationDone, 0, waitEndTime - now);
       done = emscripten_atomic_load_u32(&call->operationDone);
+      if (!done && !notifiedThread) {
+        // If we are not done even after the wait, notify the target thread,
+        // which in a race condition may have finished handling its event queue
+        // just after we added our event. (We could also notify it once right
```

Comment:
How is this possible? It looks like all enqueue and dequeue operations are protected by the `call_queue_lock`, so the event must have been added either after the target thread finished handling its event queue and released the lock or before the target thread finished handling its event queue, in which case the event would be handled because the enqueue would synchronize with the dequeue.

Reply:
I'm not entirely sure. But we get clear deadlocks without this patch, where the main thread needs to be woken up, which this patch fixes (see #12258).

It may be that there is something not entirely atomic about our mutexes, in which case I'm not sure what the best debugging approach is (maybe we need to debug the browser itself?).

Reply:
> It may be that there is something not entirely atomic about our mutexes

Yikes! Perhaps we could demonstrate such an issue with our mutexes in a smaller, controlled experiment? A test of a dining philosophers solution could be a good simple stress test for deadlock. It would also be good to narrow down whether the lock misbehaves on the main thread or on non-main threads.

Reply:
#12258 has the smallest controlled experiment I can get so far. But it still depends on allocation, proxying, and mutexes...